cmd/infra/azure/destroy.go

```diff
@@ -126,50 +127,61 @@ func (o *DestroyInfraOptions) GetResourceGroupName() string {
 	return o.Name + "-" + o.InfraID
 }
 
-func destroyServicePrincipals(o *DestroyInfraOptions) error {
-	clientID := o.ControlPlaneMIs.ControlPlane.ControlPlaneOperator.ClientID
-	if err := destroyServicePrincipal(clientID); err != nil {
-		return err
-	}
-	clientID = o.ControlPlaneMIs.ControlPlane.ImageRegistry.ClientID
-	if err := destroyServicePrincipal(clientID); err != nil {
-		return err
-	}
-	clientID = o.ControlPlaneMIs.ControlPlane.NodePoolManagement.ClientID
-	if err := destroyServicePrincipal(clientID); err != nil {
-		return err
-	}
-	clientID = o.ControlPlaneMIs.ControlPlane.CloudProvider.ClientID
-	if err := destroyServicePrincipal(clientID); err != nil {
-		return err
+func deleteApplicationsByInfraID(infraID string, logger logr.Logger) error {
+	prefixes := []string{
+		"azure-disk",
+		"azure-file",
+		"ciro",
+		"cloud-provider",
+		"cncc",
+		"cpo",
+		"ingress",
+		"capz",
 	}
-	clientID = o.ControlPlaneMIs.ControlPlane.Network.ClientID
-	if err := destroyServicePrincipal(clientID); err != nil {
-		return err
-	}
-	clientID = o.ControlPlaneMIs.ControlPlane.Ingress.ClientID
-	if err := destroyServicePrincipal(clientID); err != nil {
-		return err
+
+	cmdStr := `az ad app list --query "[].{name:displayName, appId:appId}" --all`
+	cmd := exec.Command("sh", "-c", cmdStr)
+	output, err := cmd.CombinedOutput()
+
+	if err != nil {
+		return fmt.Errorf("failed to retrieve applications: %w", err)
 	}
-	clientID = o.ControlPlaneMIs.ControlPlane.Disk.ClientID
-	if err := destroyServicePrincipal(clientID); err != nil {
-		return err
+
+	if strings.Contains(string(output), "ERROR") {
+		return errors.New(string(output))
 	}
-	clientID = o.ControlPlaneMIs.ControlPlane.File.ClientID
-	if err := destroyServicePrincipal(clientID); err != nil {
-		return err
+
+	var applications []Application
+	if err := json.Unmarshal(output, &applications); err != nil {
+		return fmt.Errorf("failed to parse applications: %w", err)
 	}
 
-	return nil
-}
+	var filteredApplications []Application
+	for _, app := range applications {
+		for _, prefix := range prefixes {
+			if app.Name == prefix+"-"+infraID {
+				filteredApplications = append(filteredApplications, app)
+				break
+			}
+		}
+	}
 
-func destroyServicePrincipal(clientID string) error {
-	cmdStr := `az ad app delete --id ` + clientID
-	cmd := exec.Command("sh", "-c", cmdStr)
-	output, err := cmd.CombinedOutput()
+	if len(filteredApplications) == 0 {
+		logger.Info("No service principals found to delete")
+		return nil
+	}
 
-	if err != nil {
-		return fmt.Errorf("failed to delete service principal, %s: %w", clientID, err)
+	for _, app := range filteredApplications {
+		cmdStr = `az ad app delete --id ` + app.AppID
+		cmd = exec.Command("sh", "-c", cmdStr)
+		output, err = cmd.CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("failed to delete application, %s: %w", app.AppID, err)
+		}
+		if strings.Contains(string(output), "ERROR") {
```

Comment:
If we invoke az directly (without using sh -c), then it suffices to check if the returned error is nil imo

Reply:
I like being able to see what the error was rather than just seeing and `exit status 1` it easier when it comes to debugging.